### ISSUE TITLE
Missing conversion for compiler generated implicit copy assignment operator

### ISSUE THREAD

**Issue by XLiZHI** [esbmc/esbmc#1126]
version: db4d28d4bfdbb8a80f118c15ba6fa4f5f70ad0ea

```
#include <cassert>
struct dim3
  {
    unsigned int x, y, z;
    dim3(unsigned int vx = 1, unsigned int vy = 1, unsigned int vz = 1)
    {
      x = vx;
      y = vy;
      z = vz;
    }
  };

int main() 
{ 
 dim3 blockdim(2,2,1);
 assert(blockdim.x == 2);
 dim3 xxx;
 xxx = blockdim;
 assert(xxx.x == 2);
 
 return 0;
}
```
Cmd: `esbmc main.cpp`
Result: 
```
----------------------------------------------------
Violated property:
  file main2.cpp line 19 function main
  assertion xxx.x == 2
  xxx.x == 2
```
from: https://github.com/XLiZHI/esbmc/blob/master/src/cpp/library/CUDA/call_kernel.h#L1241-L1245

When we try to use copy assignment operation, the validation succeeds but the assignment does not proceed.
goto function:
```
     1: dim3 xxx;
        // 38 no location
        FUNCTION_CALL:  dim3(&xxx, 1, 1, 1)
        // 39 file main2.cpp line 18 function main
        FUNCTION_CALL:  operator=(&xxx, blockdim)
```
Clearly, the compiler failed to automatically create the copy assignment operator struct/class/union.

The assignment succeeds when we explicitly create the operator as follows：
```
    dim3& operator=(const dim3& other) 
    {
        if (&other == this)
            return *this;

        x = other.x;
        y = other.y;
        z = other.z;

        return *this;
    }
```

**Comment by XLiZHI** [comment]
2. ~~Temporary objects~~ (Moved to https://github.com/esbmc/esbmc/issues/1176)

```
#include <cassert>
struct dim3
  {
    unsigned int x, y, z;
    dim3(unsigned int vx = 1, unsigned int vy = 1, unsigned int vz = 1)
    {
      x = vx;
      y = vy;
      z = vz;
    }
  };

int main() 
{ 
 dim3(1);
 
 return 0;
}
```
Cmd: `esbmc main.cpp`
Result:
```
----------------------------------------------------
Violated property:
  
  function call: not enough arguments


VERIFICATION FAILED
```
source:  https://github.com/XLiZHI/esbmc/blob/master/src/cpp/library/CUDA/call_kernel.h#L1241-L1245

version: db4d28d4bfdbb8a80f118c15ba6fa4f5f70ad0ea

When we create temporary objects, the constructor should be able to accept any number of arguments from **0 to 3** since we define default initial values in the struct.

But verification identification occurs when we take only one argument, goto function is as follows:

```
main (c:@F@main#):
        // 32 
        FUNCTION_CALL:  dim3(( dim3 *)1, 1, 1)
```
The verification is successful under the remaining number of parameters:
```
        FUNCTION_CALL:  dim3((unsigned int)(&tmp$1), 1, 1, 1)
        // 33 
        tmp$1;

```
It is not known why creation of temporary objects fails with only one argument.

**Comment by XLiZHI** [comment]
3. **Update:** ~~function return~~ (This issue has been resolved by https://github.com/esbmc/esbmc/pull/1138)

```
struct dim3
  {
    unsigned int x, y, z;
  };

dim3 getBlockIdx()
{
  dim3 block_index;
  return block_index;
}

int main() 
{ 
  getBlockIdx();
  return 0;
}
```
Cmd: `esbmc main.cpp`
Result:
```
----------------------------------------------------
Violated property:
  
  function call: not enough arguments


VERIFICATION FAILED
```

source: https://github.com/XLiZHI/esbmc/blob/master/src/cpp/library/CUDA/device_launch_parameters.h#L52-L90

version: db4d28d4bfdbb8a80f118c15ba6fa4f5f70ad0ea

goto function:

```
        // 36 
        FUNCTION_CALL:  return_value$_dim3$1=dim3(&block_index)
```
Most likely the correct format is as follows:

`FUNCTION_CALL:  dim3(return_value$_dim3$1, &block_index)`

**Comment by kunjsong01** [comment]
As for the first issue in : 

The frontend skipped the copy assignment due to https://github.com/esbmc/esbmc/blob/e1c267c41156a5c1cfc6356d6a2a08de5ebf0a2d/src/clang-c-frontend/clang_c_convert.cpp#L573-L574

We could make it look like: 
```cpp
if(fd.isImplicit() && !is_ConstructorOrDestructor(fd) && !fd.hasBody())
```

**Comment by mikhailramalho** [comment]
> As for the first issue in :
> 
> The frontend skipped the copy assignment due to
> 
> https://github.com/esbmc/esbmc/blob/e1c267c41156a5c1cfc6356d6a2a08de5ebf0a2d/src/clang-c-frontend/clang_c_convert.cpp#L573-L574
> 
> We could make it look like:
> 
> ```c++
> if(fd.isImplicit() && !is_ConstructorOrDestructor(fd) && !fd.hasBody())
> ```

Probably the right approach would be `if(fd.isImplicit() && !is_ConstructorOrDestructor(fd) && !fd.isCopyOperator() && fd.isMoveOperator())`, since we want to skip any implicit function/method but always parse copy/move operator (the rule of 5). 

Maybe there is a `fd.isConstructor()` now that covers all the 5 cases we are interested in?

**Comment by XLiZHI** [comment]
I think maybe we should change this in the C++ front end? Operators are not created in C.

In addition, the assignment operator is represented as an object of type clang::CXXMethodDecl, so we may need to convert `clang::FunctionDecl `to `clang::CXXMethodDecl`. For example: `methodDecl->isCopyAssignmentOperator()`.

**Comment by mikhailramalho** [comment]
I agree, it should look like:
```
bool clang_cpp_convertert::get_method(
  const clang::MethodDecl &fd,
  exprt &new_expr)
{
  // Only convert instantiated functions/methods
  if(fd.isDependentContext())
    return false;

  // Ignore implicit functions/methods, except constructors and destructors
  if(fd.isImplicit() && !is_ConstructorOrDestructor(fd) /* add missing comparisons */ )
    return false; 

  return clang_c_convertert::get_function(fd, new_expr);
}
```

**Comment by kunjsong01** [comment]
> I think maybe we should change this in the C++ front end? Operators are not created in C.
> 
> In addition, the assignment operator is represented as an object of type clang::CXXMethodDecl, so we may need to convert `clang::FunctionDecl `to `clang::CXXMethodDecl`. For example: `methodDecl->isCopyAssignmentOperator()`.

indeed, put it in clang cpp converter looks better.

**Comment by XLiZHI** [comment]
I re-read the structure of the code and found the following issues:

1.  Logically this method should be present in our CPP frontend, but since CPP we directly inherit from C and use the `clang_c_convertert::get_function(fd, new_expr)` method. The `if(fd.isImplicit() && !is_ConstructorOrDestructor(fd))` is inside the C function, so the easiest way is to modify it directly in the C frontend. Otherwise we need more changes.
https://github.com/esbmc/esbmc/compare/master...XLiZHI:esbmc:assign
Can you give me some advice?

2. When this method is added, an assertion failure occurs when a struct assignment occurs in the code:
```
esbmc: /home/lxzy/ESBMC_Project/esbmc/src/util/migrate.cpp:544: expr2tc sym_name_to_symbol(irep_idt, type2tc): Assertion `endatptr != atstr.c_str()' failed.
```
https://github.com/esbmc/esbmc/blob/94b1ae7d9b076d7928cc6a936e7a8decc2288876/src/util/migrate.cpp#L543-L544

@kunjsong01 's idea is

> The assertion failure might pertain to the use of this  for declaration reference inside the body of the copy assignment function

I've tried creating an assignment operator explicitly in structs, it works. Is it because the compiler implicitly generates a different assignment operator? I need some help understanding this behavior.

**Comment by mikhailramalho** [comment]
> I re-read the structure of the code and found the following issues:
> 
> 1. Logically this method should be present in our CPP frontend, but since CPP we directly inherit from C and use the `clang_c_convertert::get_function(fd, new_expr)` method. The `if(fd.isImplicit() && !is_ConstructorOrDestructor(fd))` is inside the C function, so the easiest way is to modify it directly in the C frontend. Otherwise we need more changes.
>    [master...XLiZHI:esbmc:assign](https://github.com/esbmc/esbmc/compare/master...XLiZHI:esbmc:assign)
>    Can you give me some advice?
> 2. When this method is added, an assertion failure occurs when a struct assignment occurs in the code:
> 
> ```
> esbmc: /home/lxzy/ESBMC_Project/esbmc/src/util/migrate.cpp:544: expr2tc sym_name_to_symbol(irep_idt, type2tc): Assertion `endatptr != atstr.c_str()' failed.
> ```
> 
> https://github.com/esbmc/esbmc/blob/94b1ae7d9b076d7928cc6a936e7a8decc2288876/src/util/migrate.cpp#L543-L544

Did you investigate why this is happening?

I see no reason why not follow my suggestion. As a matter of fact, I actually did that before. @kunjsong01 do you have the v2 that I sent you eons ago? That's how it's implemented there.

**Comment by kunjsong01** [comment]
> @kunjsong01 do you have the v2 that I sent you eons ago? 

Yep, still got it. 

@XLiZHI 
I agree with @mikhailramalho's suggestion in https://github.com/esbmc/esbmc/issues/1126#issuecomment-1629351257

**Comment by XLiZHI** [comment]
I've tried this. Maybe rewrite it as follows? ~~But I don't think that solves the problem~~.

https://github.com/esbmc/esbmc/blob/94b1ae7d9b076d7928cc6a936e7a8decc2288876/src/clang-cpp-frontend/clang_cpp_convert.cpp#L216-L233

```
bool clang_cpp_convertert::get_function(
  const clang::FunctionDecl &fd,
  exprt &new_expr)
{
  // Only convert instantiated functions/methods not depending on a template parameter
  if(fd.isDependentContext())
    return false;

  if(fd.isImplicit() && !is_ConstructorOrDestructor(fd) && !is_CopyOrMoveOperator(fd))
    return false;

  // add additional annotations for class/struct/union methods
  if(const auto *md = llvm::dyn_cast<clang::CXXMethodDecl>(&fd))
    if(annotate_class_method(md, new_expr, fd))
      return true;

  return clang_c_convertert::get_function(fd, new_expr);
}
```

~~`clang_c_convertert::get_function(fd, new_expr);` The implicit operator is still checked in this method, so we still don't get the implicit operator. Can you give me some hints?~~

**Comment by XLiZHI** [comment]
update: https://github.com/esbmc/esbmc/compare/master...XLiZHI:esbmc:assign

Changes have been made in the C++ frontend and now face the second problem:

> 2. When this method is added, an assertion failure occurs when a struct assignment occurs in the code

**Comment by mikhailramalho** [comment]
Nitpick: like my suggested code, it should be called something like `clang_cpp_convertert::get_method`, instead of `get_function`.

**Comment by XLiZHI** [comment]
I'm sorry I didn't understand you before. I changed it a little. Is it better now?

**update:** https://github.com/esbmc/esbmc/compare/master...XLiZHI:esbmc:assign

Preliminary results:
I commented the assertion and looked at the generated goto function. The section on the compiler-generated operators is as follows：
```
operator= (c:@S@dim3@F@operator=#&1$@S@dim3#):
        // 29 file main.cpp line 2 column 8 function operator=
        this->x=?0!0->x;
        // 30 file main.cpp line 2 column 8 function operator=
        this->y=?0!0->y;
        // 31 file main.cpp line 2 column 8 function operator=
        this->z=?0!0->z;
        // 32 file main.cpp line 2 column 8 function operator=
        RETURN: &(*this)
        // 33 file main.cpp line 2 column 8 function operator=
        END_FUNCTION // operator=
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I speculate that the problem may be in `?0!0`, possibly due to invalid Pointers.

more information about Implicitly-declared copy assignment operator: https://en.cppreference.com/w/cpp/language/copy_assignment

**Comment by XLiZHI** [comment]
update: 

Let's start by getting the parameters to account for the fact that they are not named:
https://github.com/esbmc/esbmc/blob/39458a3f4e39d2e8d1a594ec93a02856b6388c1b/src/clang-cpp-frontend/clang_cpp_convert.cpp#L997-L1002
I found the method and added a new case, when the argument is unnamed and the default assignment operator:
```
  if(is_CopyOrMoveOperator(dcxt) && is_defaulted_op(dcxt))
  {
    get_op_name(
      static_cast<const clang::CXXMethodDecl *>(dcxt), id, name, param);
    return true;
  }
```
I mimicked the code above that handles the default copy constructor, name the parameter 'ref'.
```
(gdb) param.dump()
  * type: pointer
      * subtype: symbol
          * identifier: tag-dim3
          * #constant: 1
      * #reference: 1
  * #base_name: ref
```
I still get the same validation result as above.

Unfortunately, this doesn't solve the problem. Did I do something wrong?
